File: Easy/1260_Shift_2d_grid.cpp
///first attempt 41.67%faster, 100%less memory
//Idea:Move first k mod (size_of_grid) elements to the back
class Solution {
public:
    vector<vector<int>> shiftGrid(vector<vector<int>>& grid, int k) {
    vector<int>intarray;
    vector<vector<int>>returnarray;
    for(int i=0;i<grid.size();i++)//convert 2D array into 1D
    {
        for(int j=0;j<grid[i].size();j++)
        {
            intarray.push_back(grid[i][j]);
        }
    }
    int p=intarray.size()-1;
    int m=k%(grid.size()*grid[0].size());//take mod
    vector<int>lastpart;
    while(m!=0)//save first k mod (girdsize) elements into a temp vector
    {
        lastpart.push_back(intarray[p]);
        p--;
        m--;
    }
    
    int q=0;
    int r=lastpart.size()-1;
    vector<int>temp;
    for(int i=0;i<grid.size();i++)//convert 1D array again into 2D, insert the temp vector first and then the rest
    {
        for(int j=0;j<grid[i].size();j++)
        {
            if(r!=-1)
            {
                temp.push_back(lastpart[r]);
                
                r--;
                
            }
            else
            {
                temp.push_back(intarray[q]);
                q++;
            }
        }
        returnarray.push_back(temp);
        temp.clear();

    }
    return returnarray;
    }
};

File: Easy_python/1260_shift_2D_grid.py
class Solution:
    def shiftGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:
        if k==0:
            return grid
        returnlist=[[0]*len(grid[0]) for i in range(len(grid))]
        temparr=[]
        k=k%(len(grid[0])*len(grid))
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                temparr.append(grid[i][j])
        temparr=temparr[-k:]+temparr[:-k]
        t=0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                returnlist[i][j]=temparr[t]
                t=t+1
        return returnlist
